File: proccontrol/pumpctrl.cpp
// g++ -o pumpctrl pumpctrl.cpp

#include <thread>
#include <chrono>
#include <cmath>
#include <ctime>
#include <iostream>
#include <cstddef>
#include <cstdio>

#include "pumpctrl.h"

void PumpDataProcess::add_pump_step(PumpDataProcess_t* pdp, double flowrate_ul_min, double vol_ul)
{
	PumpStep_t step;

	double flowrate_per_sec = flowrate_ul_min / 60.0;
	double duration_s = std::abs(vol_ul / flowrate_per_sec);
	double turns = std::abs(vol_ul / pdp->vol_per_turn_ul);
	double turn_per_sec = turns / duration_s;

	step.duration_s = duration_s;
	step.turn_per_sec = turn_per_sec;

	(pdp->pumpSteps).push_back(step);
}

std::size_t PumpDataProcess::get_pump_steps_size(PumpDataProcess_t* pdp)
{
	return pdp->pumpSteps.size();
}

std::vector<PumpDataProcess::PumpStep_t> PumpDataProcess::get_pump_steps(PumpDataProcess_t* pdp)
{
	return pdp->pumpSteps;
}

PumpDataProcess::Routine_t PumpDataProcess::load_into_routine(PumpDataProcess_t* pdp,
															  int iterations)
{
	Routine_t routine;
	routine.iterations = iterations;

	if ((routine.iterations == 1) || ((pdp->pumpSteps).size() % 2 == 0)) {
		routine.step_idx = -1;
		if ((pdp->pumpSteps).size() == 1) {
			routine.step_idx = -1;
		}
	}

	if (iterations > 1 && (pdp->pumpSteps).size() % 2 != 0 && (pdp->pumpSteps).size() > 1) {
		routine.step_idx = routine.iterations - 2;
	}

	if ((iterations > 1) && (pdp->pumpSteps).size() == 1) {
		routine.step_idx = routine.iterations - 2;
	}

	routine.steps = pdp->pumpSteps;
	return routine;
}

bool PumpDataProcess::isDone(Routine_t* r)
{
	if (r->steps.empty()) {
		printf("Steps empty!");
		return true;
	}
	if (r->iterations == -1) {
		printf("Iterations equal to -1");
		return false;
	}

	int iterations = r->iterations;
	printf("Total iterations: %d\n", iterations);
	bool iterationsDone = iterationsCompleted(r) >= iterations;
	return iterationsDone;
}

int PumpDataProcess::iterationsCompleted(Routine_t* r)
{
	printf("Before calculation: step_idx=%d, steps.size()=%zu, iterations=%d\n", r->step_idx,
		   r->steps.size(), r->iterations);

	int iterations_completed;

	if (r->steps.size() % 2 == 0) {
		iterations_completed = r->step_idx / r->steps.size();
	} else {
		iterations_completed = (r->step_idx + 1) / (r->steps.size() + 1);
	}

	printf("Iterations Completed: %d out of %d\n", iterations_completed, r->iterations);

	return iterations_completed;
}

void PumpDataProcess::setTimer(Routine_t* r, double duration)
{
	r->timer_duration = duration;
	r->start_time = static_cast<double>(std::time(nullptr));

	std::cout << "Timer set. Duration: " << duration << " seconds. Start Time: " << r->start_time
			  << std::endl;
}

double PumpDataProcess::elapsed(Routine_t* r)
{
	return static_cast<double>(std::time(nullptr)) - r->start_time;
}

double PumpDataProcess::untilNext(Routine_t* r)
{
	double time = r->timer_duration - elapsed(r);
	printf("Time until next action: %f seconds\n", (time < 0) ? 0.0 : time);
	return (time < 0) ? 0.0 : time;
}

bool PumpDataProcess::newActionDue(Routine_t* r)
{
	return untilNext(r) <= 0;
}

PumpDataProcess::PumpStep_t PumpDataProcess::schedActionOrNone(Routine_t* r)
{
	if (!newActionDue(r) || isDone(r)) {
		return {};
	}

	printf("__________________ step_idx incremented! __________________\n");
	r->step_idx += 1;
	if (r->step_idx != 0 && r->step_idx % r->steps.size() == 0) {
		if (iterationsCompleted(r) > r->iterations) {
			return {};
		}
	}

	PumpStep_t action = r->steps[r->step_idx % r->steps.size()];
	setTimer(r, action.duration_s);
	return action;
}

int main()
{
	PumpDataProcess pumpProcess;

	PumpDataProcess::PumpDataProcess_t pumpData;
	pumpData.vol_per_turn_ul = 10.0;

	pumpProcess.add_pump_step(&pumpData, 1200.0, 40.0);
	pumpProcess.add_pump_step(&pumpData, 1200.0, 30.0);
	pumpProcess.add_pump_step(&pumpData, 1200.0, 30.0);

	std::vector<PumpDataProcess::PumpStep_t> pumpSteps = pumpProcess.get_pump_steps(&pumpData);

	for (const auto& step : pumpSteps) {
		std::cout << "Duration: " << step.duration_s
				  << " seconds, Turn per sec: " << step.turn_per_sec << std::endl;
	}

	std::size_t pumpStepsSize = pumpProcess.get_pump_steps_size(&pumpData);
	std::cout << "Size of pumpSteps vector: " << pumpStepsSize << std::endl;

	int iterations = 3;
	PumpDataProcess::Routine_t routine = pumpProcess.load_into_routine(&pumpData, iterations);

	std::cout << "Routine Steps:" << std::endl;
	for (const auto& step : routine.steps) {
		std::cout << "Duration: " << step.duration_s
				  << " seconds, Turn per sec: " << step.turn_per_sec << std::endl;
	}

	pumpProcess.iterationsCompleted(&routine);

	while (!pumpProcess.isDone(&routine)) {
		std::cout << "Looping..." << std::endl;
		PumpDataProcess::PumpStep_t action = pumpProcess.schedActionOrNone(&routine);

		if (action.duration_s > 0) {
			std::cout << "Performing action with duration: " << action.duration_s << " seconds\n";
		}

		std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(1000)));
	}

	return 0;
}

File: proccontrol/pumpctrl.py
import time


class PumpStep:
    def __init__(self, duration_s: float = 0.0, turn_per_sec: float = 0.0):
        self.duration_s = duration_s
        self.turn_per_sec = turn_per_sec

    def string(self) -> str:
        return f"{self.duration_s:.1f}s @ {self.turn_per_sec:.5f}T/s"

    def _dirFromSpeed(self, val: float) -> bool:
        return val < 0

    def dir(self) -> bool:
        """returns the "direction" of the pump. True means "CCW" in modbus terms."""
        return self._dirFromSpeed(self.turn_per_sec)

    def turns(self) -> float:
        """Returns turns made in action."""
        return (
            self.duration_s * self.turn_per_sec if self.turn_per_sec != 0 else 0.0
        )  # s * (T/s)

    def rpm(self) -> float:
        """Returns rotations per minute."""
        return self.turn_per_sec * 60


class PumpDataProcess:
    def __init__(self, vol_per_turn_ul: float):
        self.vol_per_turn_ul = vol_per_turn_ul

    def create_pump_step(self, flowrate_ul_min: float, vol_ul: float):
        flowrate_per_sec = flowrate_ul_min / 60
        duration_s = abs(vol_ul / flowrate_per_sec)
        turns = vol_ul / self.vol_per_turn_ul
        turn_per_sec = turns / duration_s
        step = PumpStep(duration_s, turn_per_sec)
        return step

    def gen_pump_steps_list(self, *steps):
        recipe = [
            self.create_pump_step(flowrate_ul_min, vol_ul)
            for flowrate_ul_min, vol_ul in steps
        ]

        # Print the contents of the pump_steps_list
        print("\nPump Steps List:")
        for i, step in enumerate(recipe, start=1):
            print(
                f"Step {i} - Duration: {step.duration_s:.2f} seconds, Turn Rate: {step.turn_per_sec:.2f} turns per second"
            )

        return recipe

    def interpretStep(self, step: PumpStep) -> tuple[float, float]:
        """returns the step as total uL transferred (with sign) and flowrate in uL/min."""
        if step.turn_per_sec == 0:
            return 0.0, 0.0

        vol_per_turn = self.vol_per_turn_ul
        vol = vol_per_turn * step.turns()  # [uL/T] * [T] = uL
        flowrate = step.rpm() * vol_per_turn  # [T/min] * [uL/T] = uL / min
        return vol, flowrate


class Routine:
    def __init__(self, pump_steps_list: list[PumpStep] = None, iterations: int = 0):
        self.steps = pump_steps_list
        self.iterations = iterations
        self.step_idx = -1
        self.start_time = 0.0
        self.timer_duration = 0.0

    def iterationsCompleted(self):
        return self.step_idx // len(self.steps)

    def currentIndex(self) -> int:
        return self.step_idx % len(self.steps)

    def len(self) -> int:
        return len(self.steps)

    def currentAction(self) -> PumpStep | None:
        if self.isDone() or self.step_idx == -1:  # Done or not yet started.
            return None
        return self.steps[self.step_idx % len(self.steps)]

    def schedActionOrNone(self) -> PumpStep | None:
        """Returns new action if it is time to start new action, otherwise None."""
        if not self.newActionDue() or self.isDone():
            return None

        self.step_idx += 1
        if self.step_idx != 0 and self.step_idx % len(self.steps) == 0:
            if self.iterationsCompleted() == self.iterations:
                return None  # Done!

        action = self.currentAction()
        self.setTimer(action.duration_s)
        return action

    def isDone(self):
        """Returns True if routine is done."""
        if self.steps is None or len(self.steps) == 0:
            return True  # Early return for no steps found.
        if self.iterations == -1:
            return False  # Early return for infinite iterations.
        iterationsDone = self.iterationsCompleted() >= self.iterations
        # We are not done until current action is done.
        return iterationsDone

    def setTimer(self, duration: float):
        self.start_time = time.time()
        self.timer_duration = duration

    def elapsed(self) -> float:
        """Returns amount of time elapsed since start of current action."""
        return time.time() - self.start_time

    def untilNext(self) -> float:
        """Returns the time in seconds until next action is due.
        Will return a negative number if current action is done."""
        return self.timer_duration - self.elapsed()

    def newActionDue(self):
        """
        Returns True if it is time to start new action with call to update.
        """
        return self.untilNext() <= 0


class RoutineProcessor:
    def __init__(self, routine: Routine, id: int):
        self.r = routine
        self.id = id

    def update(self):
        finished = self.r.isDone()
        print(f"Finish state of routine number {self.id}: ", finished)
        if not finished:
            action = self.r.schedActionOrNone()
            if action == None:
                return
            print(f"New action! {action.string()} until-next={self.r.untilNext():.1f}s")
            vol, flowrate = pdp.interpretStep(action)
            print(
                f"sequences completed: {self.r.iterationsCompleted()}, index: {self.r.currentIndex() + 1} / {self.r.len()}, "
                f"vol: {vol:.0f}uL, flowrate: {flowrate:.1f}uL/min, turns: {action.turns():.3f}, "
                f"velocity: {action.turn_per_sec:.1e}T/s, untilnext: {self.r.untilNext():.1f}s"
            )


pump1_flag = True
pump2_flag = False
pump3_flag = False

# test
if __name__ == "__main__":
    pdp = PumpDataProcess(vol_per_turn_ul=100)
    if pump1_flag:
        pump_steps_list1 = pdp.gen_pump_steps_list(
            (4000, 200),  # flowrate_ul_min, vol_ul
            (4000, 200),
            (4000, 200),
        )
        r1 = Routine(pump_steps_list1, iterations=1)
        pr1 = RoutineProcessor(r1, 1)

    if pump2_flag:
        pump_steps_list2 = pdp.gen_pump_steps_list(
            (4000, 400),  # flowrate_ul_min, vol_ul
            (4000, 400),
            (4000, 400),
        )
        r2 = Routine(pump_steps_list2, iterations=1)
        pr2 = RoutineProcessor(r2, 2)

    if pump3_flag:
        pump_steps_list3 = pdp.gen_pump_steps_list(
            (4000, 600),  # flowrate_ul_min, vol_ul
            (4000, 600),
            (4000, 600),
        )
        r3 = Routine(pump_steps_list3, iterations=1)
        pr3 = RoutineProcessor(r3, 2)

    while True:
        if pump1_flag:
            pr1.update()
        if pump2_flag:
            pr2.update()
        if pump3_flag:
            pr3.update()
        time.sleep(1)